src/lib/__tests__/InstantSearch-test.js

```diff
@@ -1633,6 +1633,58 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/instantsear
     });
   });
 
+  test('notifies all middlewares in multi-index when called multiple times', () => {
+    const searchClient = createSearchClient();
+    const search = new InstantSearch({
+      indexName: 'indexName',
+      searchClient,
+    });
+
+    search.addWidgets([
+      connectSearchBox(() => {})({}),
+      index({ indexName: 'test' }),
+    ]);
+
+    const onMiddlewareStateChange = jest.fn();
+    const middleware = () => {
+      return {
+        subscribe() {},
+        unsubscribe() {},
+        onStateChange: onMiddlewareStateChange,
+      };
+    };
+
+    search.EXPERIMENTAL_use(middleware);
+    search.start();
+    expect(onMiddlewareStateChange).toHaveBeenCalledTimes(0);
+
+    search.setUiState({
+      indexName: {},
+      test: {},
+    });
+
+    expect(onMiddlewareStateChange).toHaveBeenCalledTimes(2);
```

Comment:
called once per index instead of only once (before), I don't see an easy way to change that tbh

Reply:
ah, it was called twice because there was an index widget in the tree. but I fear calling onMiddlewareStateChange once per index might lead to a weird interim state especially for routing. just a hunch/concern, not a concrete observation, though.

Reply:
solved by deferring the state change, which causes the onStateChange only to be called once